File: backends/python-fastapi/app/application/services/totp_service.py
from app.infrastructure.auth.totp import (
    generate_totp_secret,
    generate_totp_uri,
    verify_totp_code,
    generate_backup_codes,
    hash_backup_code,
)


def setup_totp(email: str) -> dict:
    secret = generate_totp_secret()
    uri = generate_totp_uri(secret, email)
    backup_codes = generate_backup_codes()
    hashed_codes = [hash_backup_code(c) for c in backup_codes]
    return {
        "secret": secret,
        "qr_code_uri": uri,
        "backup_codes": backup_codes,
        "hashed_backup_codes": hashed_codes,
    }


def verify_totp(secret: str, code: str) -> bool:
    return verify_totp_code(secret, code)

File: backends/go-gin/internal/application/totp_service.go
package application

import "url-shortener-go/internal/infrastructure"

type TotpSetupResult struct {
	Secret            string
	QRCodeURI         string
	BackupCodes       []string
	HashedBackupCodes []string
}

func SetupTOTP(email string) *TotpSetupResult {
	secret := infrastructure.GenerateTOTPSecret()
	uri := infrastructure.GenerateTOTPURI(secret, email, "URL Shortener")
	codes := infrastructure.GenerateBackupCodes(8)
	hashed := make([]string, len(codes))
	for i, c := range codes {
		hashed[i] = infrastructure.HashBackupCode(c)
	}

	return &TotpSetupResult{
		Secret:            secret,
		QRCodeURI:         uri,
		BackupCodes:       codes,
		HashedBackupCodes: hashed,
	}
}

func VerifyTOTP(secret, code string) bool {
	return infrastructure.VerifyTOTPCode(secret, code)
}
